File: frontend/src/types/analytics.ts
/**
 * Analytics types for admin dashboard
 * Feature: 021-admin-analytics-menu
 */

/** Период для фильтрации данных аналитики */
export type AnalyticsPeriod = '7d' | '30d' | '90d' | 'all';

/** Интервал агрегации для истории слушателей */
export type HistoryInterval = 'hour' | 'day';

/** Статистика слушателей */
export interface ListenerStats {
  /** Текущее количество слушателей */
  current: number;
  /** Пиковое значение за сегодня */
  peak_today: number;
  /** Пиковое значение за неделю */
  peak_week: number;
  /** Среднее за неделю */
  average_week: number;
}

/** Точка на графике истории слушателей */
export interface ListenerHistoryPoint {
  /** Временная метка */
  timestamp: string;
  /** Количество слушателей */
  count: number;
}

/** Ответ с историей слушателей */
export interface ListenerHistoryResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Данные для графика */
  data: ListenerHistoryPoint[];
}

/** Ответ со статистикой слушателей */
export interface ListenerStatsResponse extends ListenerStats {
  /** Кэшировано в */
  cached_at?: string;
}

/** Трек в топе */
export interface TopTrackItem {
  /** ID трека */
  track_id: number;
  /** Название трека */
  title: string;
  /** Исполнитель */
  artist: string | null;
  /** Количество воспроизведений */
  play_count: number;
  /** Общая длительность воспроизведения в секундах */
  total_duration_seconds: number;
}

/** Ответ с топом треков */
export interface TopTracksResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Список треков */
  tracks: TopTrackItem[];
}

/** Сводная статистика */
export interface AnalyticsSummaryResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Общее количество воспроизведений */
  total_plays: number;
  /** Общее время вещания в часах */
  total_duration_hours: number;
  /** Количество уникальных треков */
  unique_tracks: number;
  /** Статистика слушателей */
  listeners: ListenerStats;
  /** Время кэширования */
  cached_at: string;
}

/** Запрос на запись воспроизведения трека (internal) */
export interface TrackPlayRequest {
  /** ID трека */
  track_id: number;
  /** Длительность в секундах */
  duration_seconds?: number;
  /** Количество слушателей */
  listeners_count: number;
}

/** Ответ на запись воспроизведения трека */
export interface TrackPlayResponse {
  /** ID записи */
  id: number;
  /** Время записи */
  played_at: string;
}

// === Engagement Metrics Types ===

/** Точка на графике вовлеченности */
export interface EngagementTrendPoint {
  /** Временная метка */
  timestamp: string;
  /** Количество сообщений в чате */
  message_count: number;
  /** Количество реакций */
  reaction_count: number;
  /** Количество уникальных пользователей */
  unique_users: number;
}

/** Активный пользователь в рейтинге */
export interface ActiveUserItem {
  /** ID пользователя */
  user_id: number | null;
  /** Имя пользователя */
  username: string | null;
  /** Количество сообщений */
  message_count: number;
  /** Количество реакций */
  reaction_count: number;
  /** Время последней активности */
  last_activity: string;
}

/** Метрики вовлеченности за период */
export interface EngagementMetricsResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Общее количество сообщений */
  total_messages: number;
  /** Общее количество реакций */
  total_reactions: number;
  /** Общее количество комментариев */
  total_comments: number;
  /** Количество уникальных пользователей */
  unique_users: number;
  /** Среднее количество событий в день */
  average_daily: number;
  /** Топ активных пользователей */
  top_active_users: ActiveUserItem[];
  /** Данные для графика */
  engagement_over_time: EngagementTrendPoint[];
  /** Время кэширования */
  cached_at: string;
}

// === Stream Performance Types ===

/** Распределение по качеству */
export interface QualityDistributionItem {
  /** Уровень качества */
  quality: string;
  /** Количество записей */
  count: number;
  /** Процент от общего количества */
  percentage: number;
}

/** Точка на графике качества потока */
export interface QualityTrendPoint {
  /** Временная метка */
  timestamp: string;
  /** Общее качество */
  overall_quality: string;
  /** Аудио битрейт */
  audio_bitrate_kbps: number | null;
  /** Видео битрейт */
  video_bitrate_kbps: number | null;
  /** Процент буферизации */
  buffering_percentage: number | null;
}

/** Показатели производительности потока */
export interface StreamPerformanceResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Процент аптайма */
  uptime_percentage: number;
  /** Аптайм в часах */
  uptime_hours: number;
  /** Средний процент буферизации */
  average_buffering_percentage: number;
  /** Количество изменений качества */
  quality_changes_count: number;
  /** Использование полосы пропускания */
  bandwidth_usage_mbps: number | null;
  /** Текущее качество потока */
  current_quality: string;
  /** Распределение по качеству */
  quality_distribution: QualityDistributionItem[];
  /** Данные для графика */
  quality_over_time: QualityTrendPoint[];
  /** Время кэширования */
  cached_at: string;
}

// === Content Insights Types ===

/** Элемент контента в рейтинге */
export interface ContentPerformanceItem {
  /** ID контента */
  content_id: string;
  /** Название контента */
  title: string;
  /** Общее количество просмотров */
  total_views: number;
  /** Средний процент досмотра */
  average_completion_percentage: number;
  /** Общее время просмотра в минутах */
  total_watch_time_minutes: number;
  /** Средняя длительность просмотра */
  average_watch_duration_seconds: number;
}

/** Точка отказа (drop-off point) */
export interface DropOffPoint {
  /** Позиция в секундах */
  position_seconds: number;
  /** Процент зрителей, остановившихся здесь */
  percentage: number;
  /** Количество зрителей */
  viewers_count: number;
  /** Кумулятивный процент отказа */
  cumulative_drop_off: number;
}

/** Аналитика контента */
export interface ContentInsightsResponse {
  /** Период данных */
  period: AnalyticsPeriod;
  /** Самый просматриваемый контент */
  most_watched: ContentPerformanceItem[];
  /** Точки отказа */
  drop_off_points: DropOffPoint[];
  /** Средний рейтинг завершения */
  average_completion_rate: number;
  /** Общее количество сессий */
  total_sessions: number;
  /** Средняя длительность сессии */
  average_session_duration_seconds: number;
  /** Время кэширования */
  cached_at: string;
}

File: backend/src/schemas/analytics.py
"""
Analytics API schemas
Feature: 021-admin-analytics-menu
"""
from datetime import datetime, timezone
from typing import List, Literal, Optional
from pydantic import BaseModel, Field


# === Types ===

AnalyticsPeriod = Literal["7d", "30d", "90d", "all"]
HistoryInterval = Literal["hour", "day"]
InteractionPeriod = Literal["1h", "24h", "7d", "30d", "90d", "all"]


# === Listener Schemas ===

class ListenerStatsResponse(BaseModel):
    """Текущая статистика слушателей."""
    current: int = Field(..., description="Текущее количество слушателей")
    peak_today: int = Field(..., description="Пиковое значение за сегодня")
    peak_week: int = Field(..., description="Пиковое значение за неделю")
    average_week: float = Field(..., description="Среднее за неделю")


class ListenerHistoryPoint(BaseModel):
    """Точка на графике истории слушателей."""
    timestamp: datetime = Field(..., description="Временная метка")
    count: int = Field(..., ge=0, description="Количество слушателей")


class ListenerHistoryResponse(BaseModel):
    """История слушателей за период."""
    period: AnalyticsPeriod = Field(..., description="Период данных")
    data: List[ListenerHistoryPoint] = Field(default_factory=list, description="Данные для графика")


# === Top Tracks Schemas ===

class TopTrackItem(BaseModel):
    """Трек в топе."""
    track_id: int = Field(..., description="ID трека")
    title: str = Field(..., description="Название трека")
    artist: Optional[str] = Field(None, description="Исполнитель")
    play_count: int = Field(..., ge=0, description="Количество воспроизведений")
    total_duration_seconds: int = Field(..., ge=0, description="Общая длительность в секундах")


class TopTracksResponse(BaseModel):
    """Топ треков за период."""
    period: AnalyticsPeriod = Field(..., description="Период данных")
    tracks: List[TopTrackItem] = Field(default_factory=list, description="Список треков")


# === Summary Schema ===

class AnalyticsSummaryResponse(BaseModel):
    """Сводная статистика."""
    period: AnalyticsPeriod = Field(..., description="Период данных")
    total_plays: int = Field(..., ge=0, description="Общее количество воспроизведений")
    total_duration_hours: float = Field(..., ge=0, description="Общее время вещания в часах")
    unique_tracks: int = Field(..., ge=0, description="Количество уникальных треков")
    listeners: ListenerStatsResponse = Field(..., description="Статистика слушателей")
    cached_at: datetime = Field(..., description="Время кэширования")


# === Internal Schemas (for streamer) ===

class TrackPlayRequest(BaseModel):
    """Запрос на запись воспроизведения трека."""
    track_id: int = Field(..., description="ID трека")
    duration_seconds: Optional[int] = Field(None, ge=0, description="Длительность в секундах")
    listeners_count: int = Field(..., ge=0, description="Количество слушателей")


class TrackPlayResponse(BaseModel):
    """Ответ на запись воспроизведения трека."""
    id: int = Field(..., description="ID записи")
    played_at: datetime = Field(..., description="Время записи")


# === Interaction Analytics Schemas ===

class MostVotedPoll(BaseModel):
    """Самый популярный опрос."""
    id: str = Field(..., description="ID опроса")
    question: str = Field(..., description="Вопрос опроса")
    vote_count: int = Field(..., ge=0, description="Количество голосов")


class PollStatsResponse(BaseModel):
    """Статистика опросов."""
    total_polls: int = Field(..., ge=0, description="Всего опросов")
    active_polls: int = Field(..., ge=0, description="Активных опросов")
    total_votes: int = Field(..., ge=0, description="Всего голосов")
    unique_voters: int = Field(..., ge=0, description="Уникальных проголосовавших")
    avg_participation_rate: float = Field(..., ge=0, description="Средний уровень участия")
    most_voted_poll: Optional[MostVotedPoll] = Field(None, description="Самый популярный опрос")


class QAStatsResponse(BaseModel):
    """Статистика Q&A."""
    total_questions: int = Field(..., ge=0, description="Всего вопросов")
    pending_questions: int = Field(..., ge=0, description="Ожидающих ответа")
    answered_questions: int = Field(..., ge=0, description="Отвеченных вопросов")
    total_upvotes: int = Field(..., ge=0, description="Всего upvotes")
    unique_participants: int = Field(..., ge=0, description="Уникальных участников")
    avg_answer_time_hours: Optional[float] = Field(None, description="Среднее время ответа (часы)")


class EmojiUsage(BaseModel):
    """Использование эмодзи."""
    emoji: str = Field(..., description="Эмодзи")
    count: int = Field(..., ge=0, description="Количество использований")


class ReactionStatsResponse(BaseModel):
    """Статистика реакций."""
    total_reactions: int = Field(..., ge=0, description="Всего реакций")
    unique_users: int = Field(..., ge=0, description="Уникальных пользователей")
    top_emojis: List[EmojiUsage] = Field(default_factory=list, description="Топ эмодзи")
    reactions_per_hour: float = Field(..., ge=0, description="Реакций в час")


class ChatStatsResponse(BaseModel):
    """Статистика чата."""
    total_messages: int = Field(..., ge=0, description="Всего сообщений")
    unique_authors: int = Field(..., ge=0, description="Уникальных авторов")
    avg_message_length: float = Field(..., ge=0, description="Средняя длина сообщения")
    messages_per_hour: float = Field(..., ge=0, description="Сообщений в час")
    filtered_messages: int = Field(..., ge=0, description="Отфильтрованных сообщений")


class ActiveUser(BaseModel):
    """Активный пользователь."""
    user_id: int = Field(..., description="ID пользователя")
    username: str = Field(..., description="Имя пользователя")
    interaction_count: int = Field(..., ge=0, description="Количество взаимодействий")


class EngagementSummaryResponse(BaseModel):
    """Сводная статистика взаимодействий."""
    total_interactions: int = Field(..., ge=0, description="Всего взаимодействий")
    poll_participation_rate: float = Field(..., ge=0, le=100, description="Участие в опросах (0-100)")
    qa_engagement_rate: float = Field(..., ge=0, le=100, description="Вовлеченность в Q&A (0-100)")
    reaction_intensity: float = Field(..., ge=0, le=100, description="Интенсивность реакций (0-100)")
    chat_activity_level: float = Field(..., ge=0, le=100, description="Активность чата (0-100)")
    most_active_users: List[ActiveUser] = Field(default_factory=list, description="Самые активные пользователи")
    peak_interaction_hour: Optional[str] = Field(None, description="Пиковый час взаимодействий")


class InteractionMetricsResponse(BaseModel):
    """Полные метрики взаимодействий."""
    period: InteractionPeriod = Field(..., description="Период данных")
    polls: PollStatsResponse = Field(..., description="Статистика опросов")
    qa: QAStatsResponse = Field(..., description="Статистика Q&A")
    reactions: ReactionStatsResponse = Field(..., description="Статистика реакций")
    chat: ChatStatsResponse = Field(..., description="Статистика чата")
    engagement: EngagementSummaryResponse = Field(..., description="Сводная статистика")
    cached_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), description="Время кэширования")
